stock_putaway_storage_type_strategy/models/stock_location.py

```diff
@@ -0,0 +1,160 @@
+        for location in self:
+            if (
+                location._package_storage_type_allowed(
+                    package_storage_type, product=product
+                )
+            ):
+                return location
+        return self.browse()
+
+    def _get_ordered_children_locations(self):
+        return self.search(
+            [('id', 'child_of', self.ids), ('id', '!=', self.id)]
+        )
+
+    def _package_storage_type_allowed(self, package_storage_type, product=None):
+        self.ensure_one()
+        matching_location_storage_types = self.allowed_stock_location_storage_type_ids.filtered(
+            lambda slst: package_storage_type in slst.stock_package_storage_type_ids
+        )
+        allowed_location_storage_types = self.filter_restrictions(
+            matching_location_storage_types, product=product
+        )
+        return not self.allowed_stock_location_storage_type_ids or allowed_location_storage_types
+
+    def filter_restrictions(self, matching_location_storage_types, product=None):
+        allowed_location_storage_types = self.env['stock.location.storage.type']
+        for location_storage_type in matching_location_storage_types:
+            if location_storage_type.only_empty:
+                if (
+                    not self._existing_quants() and
+                    not self._existing_planned_moves()
+                ):
+                    allowed_location_storage_types |= location_storage_type
+            elif location_storage_type.do_not_mix_products:
+                if location_storage_type.do_not_mix_lots:
+                    quant = self.env.context.get('putaway_quant')
+                    lot = quant.lot_id
+                    if (
+                        not self._existing_quants(product=product, lot=lot) and
+                        not self._existing_planned_moves(
+                            product=product, lot=lot
+                        )
+                    ):
+                        allowed_location_storage_types |= location_storage_type
+                else:
+                    if (
+                        not self._existing_quants(product=product) and
+                        not self._existing_planned_moves(product=product)
+                    ):
+                        allowed_location_storage_types |= location_storage_type
+            else:
+                allowed_location_storage_types |= location_storage_type
+        return allowed_location_storage_types
+
+    def _existing_quants(self, product=None, lot=None):
+        domain = [('location_id', '=', self.id)]
+        if product is not None:
+            extra_domain = [('product_id', '!=', product.id)]
+            if lot is not None:
+                extra_domain = OR([extra_domain, [('lot_id', '!=', lot.id)]])
+            domain = AND([domain, extra_domain])
+        return self.env['stock.quant'].search(domain, limit=1)
```

Comment:
I would have built the domain in an other function, to make search(_prepare_domain_function()) and ease override